File: python/503. Next Greater Element II.py
class Solution(object):
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        st = []
        n = len(nums)
        res = [-1] * n
        for i in range(2*n):
            idx = i%n
            while st and nums[st[-1]] < nums[idx]:
                # print st, nums[st[-1]]
                res[st.pop()] = nums[idx]
            if i < n:
                st.append(i)
        return res

if __name__ == '__main__':
    nums = [2,4,1,3]
    su = Solution()
    print su.nextGreaterElements(nums)        

File: c++/503. Next Greater Element II.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n, -1);
        stack<int> st;

        for (int i = 0; i < 2*n; i++) {
            int idx = i%n;
            while (!st.empty() && nums[st.top()] < nums[idx]) {
                res[st.top()] = nums[idx];
                st.pop();
            }
            if (i < n) st.push(i);
        }
        return res;
    }
};

// 使用栈记录i之前的数字

